internal/datacoord/channel_manager_v2.go

```diff
@@ -0,0 +1,744 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datacoord
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/errors"
+	"github.com/samber/lo"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/kv"
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/conc"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+type ChannelManager interface {
+	Startup(ctx context.Context, legacyNodes, allNodes []int64) error
+	Close()
+
+	AddNode(nodeID UniqueID) error
+	DeleteNode(nodeID UniqueID) error
+	Watch(ctx context.Context, ch RWChannel) error
+	Release(nodeID UniqueID, channelName string) error
+
+	Match(nodeID UniqueID, channel string) bool
+	FindWatcher(channel string) (UniqueID, error)
+
+	GetChannel(nodeID int64, channel string) (RWChannel, bool)
+	GetNodeIDByChannelName(channel string) (int64, bool)
+	GetNodeChannelsByCollectionID(collectionID int64) map[int64][]string
+	GetChannelsByCollectionID(collectionID int64) []RWChannel
+	GetChannelNamesByCollectionID(collectionID int64) []string
+}
+
+// An interface sessionManager implments
+type SubCluster interface {
+	NotifyChannelOperation(ctx context.Context, nodeID int64, req *datapb.ChannelOperationsRequest) error
+	CheckChannelOperationProgress(ctx context.Context, nodeID int64, info *datapb.ChannelWatchInfo) (*datapb.ChannelOperationProgressResponse, error)
+}
+
+type ChannelManagerImplV2 struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	mu     sync.RWMutex
```

Comment:
This lock gurads RWChannelStore only, including all channels info in memory. It makes sure each method in channelmanager, like "AddNode" can be safe. Because ALL balances are read-write operations, locks inside RWChannelstore won't be enough